File: packages/sdk-ts/src/resources/billing.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { Balance, CreateTopup, LedgerEntry, PaginationQuery, Topup } from '../types.js';

/** Billing: the prepaid balance, its history, and adding to it. */
export class Billing {
    constructor(private readonly http: Http) {}

    /** `balance_micros`, `held_micros` (reserved by live calls) and `available_micros`. */
    balance(opts?: RequestOptions): Promise<Balance> {
        return this.http.request<Balance>('GET', '/v1/billing/balance', opts);
    }

    /** Every movement on the account. Balances reconcile exactly. */
    ledger(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<LedgerEntry>> {
        return this.http.list<LedgerEntry>('/v1/billing/ledger', query, opts);
    }

    /**
     * Start a top-up. Returns a `checkout_url` a person must open; credit
     * lands once payment settles. An `Idempotency-Key` is generated unless
     * `opts.idempotencyKey` is given.
     */
    createTopup(body: CreateTopup, opts?: RequestOptions): Promise<Topup> {
        return this.http.request<Topup>('POST', '/v1/billing/topups', { ...opts, body, idempotent: true });
    }
}

File: packages/sdk-dart/lib/src/resources/billing.dart
import '../models/billing.dart';
import '../page.dart';
import '../response.dart';
import '../transport.dart';

/// Billing: the prepaid balance, its history, and adding to it.
class Billing {
  /// Binds the resource to a transport.
  const Billing(this._transport);

  final Transport _transport;

  /// The balance, what live calls have reserved, and what is left.
  Future<Balance> balance() =>
      _transport.request('GET', '/v1/billing/balance', Balance.fromJson);

  /// Every movement on the account. Balances reconcile exactly.
  Future<Page<LedgerEntry>> ledger(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/billing/ledger', LedgerEntry.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Starts a top-up. Returns a `checkoutUrl` a person must open; credit
  /// lands once payment settles, not when this returns.
  Future<Topup> createTopup(CreateTopup body, {String? idempotencyKey}) =>
      _transport.request(
        'POST',
        '/v1/billing/topups',
        Topup.fromJson,
        body: body.toJson(),
        idempotent: true,
        idempotencyKey: idempotencyKey,
      );

  /// As [createTopup], but keeps the request id and whether the server
  /// replayed an earlier request.
  Future<WixzelResponse<Topup>> createTopupWithResponse(CreateTopup body,
          {String? idempotencyKey}) =>
      _transport.requestWithResponse(
        'POST',
        '/v1/billing/topups',
        Topup.fromJson,
        body: body.toJson(),
        idempotent: true,
        idempotencyKey: idempotencyKey,
      );
}
